src/cli/self_update/shell.rs

```diff
@@ -0,0 +1,213 @@
+//! Paths and Unix shells
+//!
+//! MacOS, Linux, FreeBSD, and many other OS model their design on Unix,
+//! so handling them is relatively consistent. But only relatively.
+//! POSIX postdates Unix by 20 years, and each "Unix-like" shell develops
+//! unique quirks over time.
+//!
+//!
+//! Windowing Managers, Desktop Environments, GUI Terminals, and PATHs
+//!
+//! Duplicating paths in PATH can cause performance issues when the OS searches
+//! the same place multiple times. Traditionally, Unix configurations have
+//! resolved this by setting up PATHs in the shell's login profile.
+//!
+//! This has its own issues. Login profiles are only intended to run once, but
+//! changing the PATH is common enough that people may run it twice. Desktop
+//! environments often choose to NOT start login shells in GUI terminals. Thus,
+//! a trend has emerged to place PATH updates in other run-commands (rc) files,
+//! leaving Rustup with few assumptions to build on for fulfilling its promise
+//! to set up PATH appropriately.
+//!
+//! Rustup addresses this by:
+//! 1) using a shell script that updates PATH if the path is not in PATH
+//! 2) sourcing this script in any known and appropriate rc file
+
+use std::path::PathBuf;
+
+use error_chain::bail;
+
+use super::*;
+use crate::process;
+
+pub type Shell = Box<dyn UnixShell>;
+
+#[derive(Debug, PartialEq)]
+pub struct ShellScript {
+    content: &'static str,
+    name: &'static str,
+}
+
+impl ShellScript {
+    pub fn write(&self) -> Result<()> {
+        let home = utils::cargo_home()?;
+        let cargo_bin = format!("{}/bin", cargo_home_str()?);
+        let env_name = home.join(self.name);
+        let env_file = self.content.replace("{cargo_bin}", &cargo_bin);
+        utils::write_file(self.name, &env_name, &env_file)?;
+        Ok(())
+    }
+}
+
+// TODO: Update into a bytestring.
+pub fn cargo_home_str() -> Result<Cow<'static, str>> {
+    let path = utils::cargo_home()?;
+
+    let default_cargo_home = utils::home_dir()
+        .unwrap_or_else(|| PathBuf::from("."))
+        .join(".cargo");
+    Ok(if default_cargo_home == path {
+        "$HOME/.cargo".into()
+    } else {
+        match path.to_str() {
+            Some(p) => p.to_owned().into(),
+            None => bail!("Non-Unicode path!"),
+        }
+    })
+}
+
+// TODO: Tcsh (BSD)
+// TODO?: Make a decision on Ion Shell, Power Shell, Nushell
+// Cross-platform non-POSIX shells have not been assessed for integration yet
+fn enumerate_shells() -> Vec<Shell> {
+    vec![Box::new(Posix), Box::new(Bash), Box::new(Zsh)]
```

Comment:
I wonder if `.xinitrc` or `.xsession` are worth updating if present?

Reply:
I don't think so but I would need to do a more extensive evaluation.

I would add X as a shell: it is in fact a user-facing environment allowing interaction with the OS :wink:, even if conventionally we limit "shell" as a term to TUIs and their easily scripted interfaces.

Reply:
Mmm, I was just thinking from the PoV of attempting to ensure the path is in the DE's environment for the purpose of things like keybinding-arranged launch windows.  But it's not important for this release, we can consider it in the future.